numpy/core/tests/test__exceptions.py

```diff
@@ -56,3 +57,32 @@ class TestUFuncNoLoopError:
     def test_pickling(self):
         """ Test that _UFuncNoLoopError can be pickled """
         assert isinstance(pickle.dumps(_UFuncNoLoopError), bytes)
+
+
+@pytest.mark.parametrize("args", [
+    (2, 1, None),
+    (2, 1, "test_prefix"),
+    ("test message",),
+])
+class TestAxisError:
+    def test_attr(self, args):
+        """Validate attribute types."""
+        exc = np.AxisError(*args)
+        if len(args) == 1:
+            assert exc.axis is None
+            assert exc.ndim is None
+        else:
+            axis, ndim, *_ = args
+            assert exc.axis == axis
+            assert exc.ndim == ndim
+
+    def test_pickling(self, args):
+        """Test that `AxisError` can be pickled."""
+        exc = np.AxisError(*args)
+        exc2 = pickle.loads(pickle.dumps(exc))
+
+        assert type(exc) is type(exc2)
+        for name in ("axis", "ndim", "args"):
+            attr1 = getattr(exc, name)
+            attr2 = getattr(exc2, name)
+            assert attr1 == attr2, name
```

Comment:
@eric-wieser there is already a similar test in place for testing pickling in general.
I'd say that this test already adequately covers pickle use cases, would you agree?

Reply:
This test is adequate for checking pickling works, but not for checking that any future changes are backwards compatible.

On the other hand, I think there is little reason to promise exceptions pickled in one version of numpy can be correctly loaded from another, so the test could be argued as unecessary.

I'll let you make the call - if we do add the test, it at least acts as a canary to prevent us breaking compatibility by accident and make us do it consciously; but it likely also makes the test fragile to changes in pickle itself which isn't our problem.

Reply:
I'm a bit wary of the additional fragility that this would add.

Previously `__reduce__` would always return the `np.AxisError` type as reconstructor and a 1-tuple with error message as parameter. As this type of signature is already tested with the last `pytest.mark.parametrize` parameter, we should be fine with leaving things as they are.